### PR TITLE
Implemented purge functionality

### PR DESCRIPTION
**Pull request by vbusch**
-Removed the inactive suspend menu item.
- Hooked up the purge menu item.

Signed-off-by: Vanessa <vbusch@redhat.com>

### Type of change

_Select the type of your PR_

- Bugfix
- Enhancement / new feature
- Refactoring
- Documentation

### Description

_Please describe your pull request_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./documentation/design`
- [ ] Write tests and make sure they pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: agent/lib/queue.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+function find_queue (name, pods) {
+    return Promise.all(pods.map(function (pod) {
+        return pod.broker.getQueueNames();
+    } )).then(function (results) {
+            var found = false;
+            var pod = undefined;
+            var min = undefined;
+            for (var i = 0; !found && i < results.length; i++) {
+                if (results[i].indexOf(name) >= 0) {
```
Comment (k-wall) [inline agent/lib/queue.js]: I query the use of `indexOf`.  if I have addresses `foo` and `foobar`, and I purge `foo`, what stops this code finding erroneously finding `foobar`?

File: agent/lib/queue.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                    pod = pods[i];
+                    found = true;
```
Comment (k-wall) [inline agent/lib/queue.js]: You are indexing the pod array (pods) with an index that relates to the queue array.  That seems wrong.

File: agent/lib/queue.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        function(currentPods) {
+
+            var podDefs = currentPods.items.map(function (pod) {
+                return pod_watcher.get_pod_definition(pod);
+            });
+            pods.update(podDefs);
+
+            return find_queue(queueName, pods.pod_list()).then(
+                function (result) {
+
```
Comment (k-wall) [inline agent/lib/queue.js]: what if a queue is shared across many brokers?  we need to purge each shard.

**Comment by lulf** [comment]
@vbusch @k-wall so thinking more about my initial comment... using status.brokerStatuses is a nice way to get all 'shards' of a queue. However, this would only work in standard address space. In brokered space, there is no such thing set in status as its just 1 broker anyway.

**Comment by k-wall** [comment]
> @vbusch @k-wall so thinking more about my initial comment... using status.brokerStatuses is a nice way to get all 'shards' of a queue. However, this would only work in standard address space. In brokered space, there is no such thing set in status as its just 1 broker anyway.

For me I don't see a problem if we take a different approach at the backend for standard vs brokered.

**Comment by vbusch** [comment]
I'll Update the way I get the broker for the standard address space.

File: agent/lib/queue.js (hunk 1 above)
Comment (vbusch) [reply agent/lib/queue.js]: results, is an array of arrays.  One for each pod.  
eg. [ ['foo', 'bar'], [queue_on_other_pod] ]
indexOf on an array, won't return foobar when searching for foo.

File: agent/lib/queue.js (hunk 2 above)
Comment (vbusch) [reply agent/lib/queue.js]: Relates to the comment above.
'i' is the index of the results for all the promises.  eg. [ ['foo', 'bar'], ['queue_on_other_pod'] ]
The 1st pod queried returned 'foo', and 'bar', and the 2nd pod queried returned 'queue_on_other_pod'.
The results are in the same order as the promises in Promise.all

File: agent/lib/queue.js (hunk 3 above)
Comment (vbusch) [reply agent/lib/queue.js]: using the address definition status, to get the brokers for standard address space.  If there are shards, it gets all of them.

File: agent/lib/queue.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                id:     podDef.name,
+            };
+        }
+        try {
+            options = tls_options.get_client_options(options);
+        } catch (error) {
+            log.error(error);
+        }
+
+        return artemis.connect(options);
```
Comment (k-wall) [inline agent/lib/queue.js]: How is this being closed?

Comment (k-wall) [reply agent/lib/queue.js]: I've used a finally callback.

File: agent/lib/queue.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                       return broker.purgeQueue(queueName);
+                   });
+               });
+
+               Promise.all(purgedQueues)
+                   .then(function (results) {
+                       if (results.length > 0) {
+                           return results[0];
+                       } else {
+                           return false;
```
[the remark this answers is not in the document]
Comment (k-wall) [reply agent/lib/queue.js]: Done

File: agent/lib/queue.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+               var brokers = create_brokers(podDefs, 'standard');
+               var purgedQueues = brokers.map(function (broker) {
+                   return (new function () {
+                       return broker.purgeQueue(queueName);
+                   });
+               });
+
+               Promise.all(purgedQueues)
+                   .then(function (results) {
+                       if (results.length > 0) {
```
[the remark this answers is not in the document]
Comment (k-wall) [reply agent/lib/queue.js]: I now sum them all.

File: agent/lib/queue.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+               });
+
+               var brokers = create_brokers(podDefs, 'standard');
+               var purgedQueues = brokers.map(function (broker) {
+                   return (new function () {
+                       return broker.purgeQueue(queueName);
+                   });
+               });
+
+               Promise.all(purgedQueues)
```
[the remark this answers is not in the document]
Comment (k-wall) [reply agent/lib/queue.js]: Done

File: agent/lib/queue.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            });
+
+            //for the brokered address space, there will never be more then 1 broker.
+            if (results.length > 0) {
+                return results[0];
+            }
+        });
+};
+
+Queue.prototype.purge = function () {
```
[the remark this answers is not in the document]
Comment (vbusch) [reply agent/lib/queue.js]: Yes, and in another PR we can log a kubernetes event.

[the remark this answers is not in the document]
Comment (k-wall) [reply agent/lib/queue.js]: Re. logging a kubernetes event for purge queue from agent is tricky.  We have an variable variable ENABLE_EVENT_LOGGER which controls whether the component logs events or not.  For standard, this is unset for agent (as standard-controller writes the event log messages), but set for brokered.  To log a purge address message for both spaces, I need to ignore this env var and log the event unconditionally.  So it is possible, but the code will be ugly.  The purge PR as it stands logs to the agent log that fact that the purge has happened, the number of messages and the userid of the invoking user.  Maybe this is sufficient for now.

**Pull request merged**